tools/migration/internal/runner.go

```diff
@@ -26,27 +41,169 @@ type Migration interface {
+	command string
+
+	// oldRepoOpt is the value of --old-repo passed  from the CLI
 	oldRepoOpt string
-}
 
-// NewMigrationRunner builds a MirgrationRunner for the given command and repo options
-func NewMigrationRunner(logger *Logger, command, oldRepoOpt string) *MigrationRunner {
-	// TODO: Issue #2585 Implement repo migration version detection and upgrade decisioning
+	// newRepoPath is where the to-be-migrated/migrated repo is located
+	newRepoPath string
 
+	// MigrationsProvider is a dependency for fetching available migrations
+	// to allow unit tests to supply test migrations without creating test fixtures.
+	MigrationsProvider func() []Migration
+}
+
+// NewMigrationRunner builds a MigrationRunner for the given command and repo options
+func NewMigrationRunner(logger *Logger, command, oldRepoOpt, newRepoOpt string) *MigrationRunner {
 	return &MigrationRunner{
-		logger:     logger,
-		command:    command,
-		oldRepoOpt: oldRepoOpt,
+		logger:             logger,
+		command:            command,
+		oldRepoOpt:         oldRepoOpt,
+		newRepoPath:        newRepoOpt,
+		MigrationsProvider: DefaultMigrationsProvider,
 	}
 }
 
 // Run executes the MigrationRunner
 func (m *MigrationRunner) Run() error {
-	// TODO: Issue #2595 Implement first repo migration
+	repoVersion, err := m.repoVersion(m.oldRepoOpt)
+	if err != nil {
+		return err
+	}
+	if repoVersion == m.getTargetMigrationVersion() {
+		return fmt.Errorf("binary version %d = repo version %d; migration not run", repoVersion, m.getTargetMigrationVersion())
+	}
+
+	var mig Migration
+	if mig, err = m.getValidMigration(repoVersion); err != nil {
+		return fmt.Errorf("migration check failed: %s", err.Error())
+	}
+	return m.runCommand(mig)
+}
+
+func (m *MigrationRunner) runCommand(mig Migration) error {
+	var err error
+
+	_, to := mig.Versions()
 
-	return m.logger.Close()
+	switch m.command {
+	case "describe":
+		m.logger.Print(mig.Describe())
+	case "migrate":
+		if m.newRepoPath, err = CloneRepo(m.oldRepoOpt); err != nil {
+			return err
+		}
+		m.logger.Print(fmt.Sprintf("new repo will be at %s", m.newRepoPath))
+
+		if err = mig.Migrate(m.newRepoPath); err != nil {
+			return errors.New("migration failed: " + err.Error())
+		}
+		if err = m.validateAndUpdateVersion(to, m.newRepoPath, mig); err != nil {
+			return errors.New("validation failed: " + err.Error())
+		}
+		if err = InstallNewRepo(m.oldRepoOpt, m.newRepoPath); err != nil {
+			return errors.New("installation failed: " + err.Error())
+		}
+	case "buildonly":
+		if m.newRepoPath, err = CloneRepo(m.oldRepoOpt); err != nil {
+			return err
+		}
+		m.logger.Print(fmt.Sprintf("new repo will be at %s", m.newRepoPath))
+
+		if err = mig.Migrate(m.newRepoPath); err != nil {
+			return errors.New("migration failed: " + err.Error())
+		}
+	case "install":
+		if m.newRepoPath == "" {
+			return errors.New("installation failed: new repo is missing")
+		}
+		if err = m.validateAndUpdateVersion(to, m.newRepoPath, mig); err != nil {
```

Comment:
IMO version bumping could fall under the actual `Migrate` function of different migrations.

Reply:
> IMO version bumping could fall under the actual Migrate function of different migrations.

I strongly feel that version bumping should be controlled by the Runner, whose job it is to also call Validation, and should not be done until Validation passes. Otherwise you leave the new repo in a half-migrated state, esp. given we use version to detect whether migration is needed.

Reply:
I forgot to add, the original reasoning behind putting validation in the Install step is so you couldn't shoot yourself in the foot if something went wrong and the failure was undetected or forgotten.  Having the install command do the validate before actually swapping out the links provides some additional safety.  Maybe there should at least be a version check before the actual Install call.